### ISSUE TITLE
Incorrect template type computed under NTI

### ISSUE THREAD

**Issue by chrisdoble** [google/closure-compiler#2022]
Please consider the following code (a contrived simplification of the code in which I found this issue):

``` js
/** @enum {string} */
const Enum = {VALUE: 'VALUE'};

/**
 * @param {string} s
 * @return {Enum}
 */
function deserializeEnum(s) {
  switch (s) {
    case 'VALUE':
      return Enum.VALUE;
    default:
      throw new Error(`invalid enum: ${s}`);
  }
}

/** @param {Enum} e */
function logEnum(e) {
  console.log(e);
}

/**
 * @template T
 * @template U
 * @param {function(U):T} f
 * @param {T} def
 * @param {U|null|undefined} val
 * @return {T}
 */
function getOrDefaultM(f, def, val) {
  return val != null ? f(val) : def;
}

const o = {e: getOrDefaultM(deserializeEnum, Enum.VALUE, '')};
logEnum(o.e);
```

Compiling this under OTI produces no warning:

```
$ java -jar compiler-latest/closure-compiler-v20160911.jar --checks_only --warning_level VERBOSE test.js
$
```

Compiling it under NTI does:

```
$ java -jar compiler-latest/closure-compiler-v20160911.jar --checks_only --new_type_inf --warning_level VERBOSE test.js
test.js:35: WARNING - Invalid type for parameter 1 of function logEnum.
Expected : Enum
Found    : string

logEnum(o.e);
        ^

0 error(s), 1 warning(s), 87.17948717948718% typed
$
```

No warning is produced under NTI if:
- the object literal is unwrapped, i.e. `const o = {e: getOrDefaultM(…)};` is replaced with `const e = getOrDefaultM(…);`, or 
- the `getOrDefaultM(…)` call is inlined into the `logEnum` call.

Is this behaviour correct? I expected the type of `o.e` to be `Enum` and no warning be produced.

**Comment by chrisdoble** [comment]
Changing `const` to `let` fixes the issue. I assume this is due to [const inference](https://github.com/google/closure-compiler/wiki/Using-NTI-(new-type-inference)#warnings-about-uninferred-constants).

**Comment by Dominator008** [comment]
@dimvar Is there anything we could improve here?

**Comment by dimvar** [comment]
@chrisdoble yes, this appears to be a bug in early const inference. The same expression has the correct type in the later type inference. [Smaller repro](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540enum%2520%257Bstring%257D%2520*%252F%250Avar%2520Enum%2520%253D%2520%257B%2520A%253A%2520'asdf'%2520%257D%253B%250A%250A%252F**%250A%2520*%2520%2540template%2520T%250A%2520*%2520%2540param%2520%257BT%257D%2520x%250A%2520*%2520%2540return%2520%257BT%257D%250A%2520*%252F%250Afunction%2520id(x)%2520%257B%2520return%2520x%253B%2520%257D%250A%250A%252F**%2520%2540const%2520*%252F%250Avar%2520ns%2520%253D%2520%257B%250A%2520%2520e%253A%2520id(Enum.A)%250A%257D%253B%250A%250Avar%2520%252F**%2520!Enum%2520*%252F%2520en%253B%250A%250Aen%2520%253D%2520ns.e%253B%2520%2520%252F%252F%2520spurious%2520warning%250A%250Aen%2520%253D%2520id(Enum.A)%253B%2520%2520%252F%252F%2520no%2520warning%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26CLOSURE_PASS%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1)
